scipy/interpolate/_rbfinterp.py

```diff
@@ -0,0 +1,557 @@
+
+    >>> import matplotlib.pyplot as plt
+    >>> from scipy.interpolate import RBFInterpolator
+    >>> from scipy.stats.qmc import Halton
+
+    >>> rng = np.random.default_rng()
+    >>> xobs = 2*Halton(2, seed=rng).random(100) - 1
+    >>> yobs = np.sum(xobs, axis=1)*np.exp(-6*np.sum(xobs**2, axis=1))
+
+    >>> xgrid = np.mgrid[-1:1:50j, -1:1:50j]
+    >>> xflat = xgrid.reshape(2, -1).T
+    >>> yflat = RBFInterpolator(xobs, yobs)(xflat)
+    >>> ygrid = yflat.reshape(50, 50)
+
+    >>> plt.pcolormesh(*xgrid, ygrid, vmin=-0.25, vmax=0.25, shading='gouraud')
+    >>> plt.scatter(*xobs.T, c=yobs, s=50, ec='k', vmin=-0.25, vmax=0.25)
+    >>> plt.colorbar()
+    >>> plt.show()
+
+    """
+
+    def __init__(self, y, d,
+                 smoothing=0.0,
+                 kernel="thin_plate_spline",
+                 epsilon=None,
+                 degree=None):
+        y, d, smoothing, kernel, epsilon, degree, _ = _sanitize_init_args(
+            y, d, smoothing, kernel, epsilon, degree, None
+            )
+
+        ny, ndim = y.shape
+        data_shape = d.shape[1:]
+        data_type = d.dtype
+        d = d.reshape((ny, -1))
+        # If `d` is complex, convert it to a float array with twice as many
+        # columns. Otherwise, the lhs matrix would need to be converted to
+        # complex and take up 2x more memory than necessary.
+        d = d.view(float)
+        powers = _monomial_powers(ndim, degree)
+        shift, scale, coeffs = _build_and_solve_system(
+            y, d, smoothing, kernel, epsilon, powers
+            )
+
+        self.y = y
+        self.kernel = kernel
+        self.epsilon = epsilon
+        self.powers = powers
+        self.shift = shift
+        self.scale = scale
+        self.coeffs = coeffs
+        self.data_shape = data_shape
+        self.data_type = data_type
+
+    def __call__(self, x):
+        """Evaluate the interpolant at `x`.
+
+        Parameters
+        ----------
+        x : (Q, N) array_like
+            Interpolation point coordinates.
+
+        Returns
+        -------
+        (Q, ...) ndarray
+            Values of the interpolant at `x`.
+
+        """
+        x = np.asarray(x, dtype=float, order="C")
+        if x.ndim != 2:
+            raise ValueError("`x` must be a 2-dimensional array.")
+
+        nx, ndim = x.shape
+        if ndim != self.y.shape[1]:
+            raise ValueError(
+                "Expected the second axis of `x` to have length "
+                f"{self.y.shape[1]}."
+                )
+
+        out = _evaluate(
+            x, self.y, self.kernel, self.epsilon, self.powers, self.shift,
+            self.scale, self.coeffs
+            )
+        out = out.view(self.data_type)
+        out = out.reshape((nx,) + self.data_shape)
+        return out
+
+
+class RBFLocalInterpolator:
+    """RBF interpolation using the k nearest data points.
+
+    See `RBFInterpolator` for more information.
```

Comment:
I am actually not seeing what you mean here. There is no example or note about this class in the other docstrings.

Reply:
I updated it to say "See the Notes section of `RBFInterpolator` for a description of RBF interpolation."

Reply:
OK but what I would be missing as a user is: why shall I use `RBFLocalInterpolator` over `RBFInterpolator`? A small description of the differences and consequences on the result would be helpful.

Reply:
I added a note saying "This class can be a faster and less memory intensive alternative to `RBFInterpolator` when interpolating many data points."

Reply:
Great better thanks. Resolving, but just out of curiosity: do you have some evidence of the benefits? Did you do some plot or something else? Because I know that KDTree can be slow in high dimensions as well.

Reply:
Here is the result of the benchmark after modifying some of the parameters. `n_samples` is both the number of observations and the number of interpolation points (which are 2-dimensional).  The runtime for `RBFLocalInterpolator` increases by O(n_samples) and the runtime for `RBFInterpolator` increases by something greater than O(n_samples^2). `KDTree` is relatively fast compared to the rest of the code. Perhaps with larger values for `n_samples` `RBFLocalInterpolator` will be something closer to O(n_samples * log(n_samples)) (or whatever it would be for a `KDTree`). I mostly use RBF interpolation for 1-3 dimensions, and I am not sure if it is still a practical tool for high dimensional interpolation. 

```
[ 75.00%] ··· interpolate.RBFInterpolator.time_rbf_interpolator                                                                                                                                           ok
[ 75.00%] ··· =========== ===================
              --                 kernel      
              ----------- -------------------
               n_samples   thin_plate_spline 
              =========== ===================
                   10         27.8±0.03μs    
                  100           521±3μs      
                  1000        57.6±0.06ms    
                 10000         17.2±0.08s    
              =========== ===================

[100.00%] ··· interpolate.RBFLocalInterpolator.time_rbf_local_interpolator                                                                                                                                ok
[100.00%] ··· =========== ===================
              --                 kernel      
              ----------- -------------------
               n_samples   thin_plate_spline 
              =========== ===================
                   10          217±0.2μs     
                  100         6.93±0.01ms    
                  1000         84.1±0.2ms    
                 10000          804±3ms      
              =========== ===================
```

Reply:
Thanks for that. Just unresolving so that people see your comment.

Reply:
Memory is a big issue with standard RBF interpolation.  More than a few 100 datapoints and you're stuck (cannot allocate the matrix that needs to be solved).  This makes RBFs useless for images; except with a workaround like this local interpolator, which greatly expands applicability.

Reply:
Ah yes I see. Then I would emphasise this more in the doc. Because I already see the bug reports coming for this otherwise 😅

Reply:
Then, could we also have a sparse version? Not based on close-by points but on a number of points in general? Some people might want to keep the general behaviour over the local one. This could be a parameter in `RBFInterpolator` which would chose `n` points from the data.

Reply:
I don't think you can have a sparse version, because each kernel contributes anywhere.  You could potentially have a truncated kernel version that is sparse by constructing coefficients in the matrix only when the kernel distances are less than a certain cut-off.

Reply:
I will add a note to `RBFInterpolator` suggesting to use `RBFLocalInterpolator` when there are more than about 1000 data points.

It is possible to do the sparse RBF interpolation like @stefanv suggested.  You would need to use the Wendland class of RBFs which have compact support.  They look a bit like Gaussian RBFs, but they reach zero at a distance `epsilon`. That gives you a sparse system to solve (assuming `epsilon` is small compared to the size of your domain). I have implemented sparse RBF interpolation with Wendland functions in my RBF package (https://github.com/treverhines/RBF). I would suggest that anyone interested should try out that implementation. If enough people are interested then maybe we can squeeze that functionality into `RBFInterpolator`, but it would be a large undertaking.